pdb: resume capturing after `continue`

blueyed opened the pull request:
After `pdb.set_trace()` capturing is turned off.
This patch resumes it after using the `continue` (or `c` / `cont`)
command.

I've noticed that any captured output from before the `set_trace` is lost, but that does not appear to be caused by my patch - need to investigate.
I've only tested this on `master` for now.

TODO:

- [x] Add a new news fragment into the changelog folder
  * name it `$issue_id.$type` for example (588.bug)
  * if you don't have an issue_id change it to the pr id after creating the pr
  * ensure type is one of `removal`, `feature`, `bugfix`, `vendor`, `doc` or `trivial`
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."
- [x] Target: for `bugfix`, `vendor`, `doc` or `trivial` fixes, target `master`; for removals or features target `features`;
- [x] Make sure to include reasonable tests for your change if necessary

blueyed commented:
> I've noticed that any captured output from before the set_trace is lost

That is: when using `import pdb; pdb.set_trace` in a test, any stderr/stdout before that gets not displayed by pytest in case the test fails later.
It can be seen when using `pytest -s`, so it should also be captured (and kept) before `set_trace` is used.

blueyed commented:
Thanks.
Pushed a fixup for also hooking into `setup`.
Don't know about a test..

Regarding the other issue (capturing being lost), I've noticed that Django's `reverse` (loading the apps / setting up Django) changes the root logger's handlers:

From `[<StreamHandler <_io.FileIO name=9 mode='rb+' closefd=True> (INFO)>, <LogCaptureHandler (NOTSET)>]` to `[<StreamHandler <stderr> (INFO)>]`.

(via `logging.getLogger().handlers`)
So this is not a pytest issue probably.

blueyed commented:
Only py27-pexpect failed, but not py35-pexpect?!

nicoddemus commented:
Hmm I'm under the impression that those "pexpect" tests are flaky; had to restart a single failing job today which passed the second time.

Retarted your `py27-pexpect` a now other tests failed.

nicoddemus commented:
@blueyed gentle ping. 👍

blueyed commented:
Rebased and fixed/adjusted.

There are two failures with py27-pexpect that need investigation:

> FAIL test_pdb.py::TestPDB::()::test_pdb_interaction_capturing_simple
FAIL test_pdb.py::TestPDB::()::test_pdb_interaction_capturing_twice

blueyed commented:
What about a new `pytest_leave_pdb` hook (to complement `pytest_enter_pdb`)?

RonnyPfannschmidt commented:
why is the hook not actually pytest_enter_pdb(config, frame), so thecapture plugin could implement a hookwrapper

i would like to propose a pytest_debug(hook) that allows integration of general debuggers

blueyed commented:
@RonnyPfannschmidt 
1. So you are suggesting to add `frame` to the existing hook, and use it internally?
2. I do not understand fully how `pytest_debug(hook)` would work.

Both points appear to be outside the scope of this PR though, but we can create new issues/PRs for them, of course.

RonnyPfannschmidt commented:
sorry my mistake, i meant to write `pytest_debug(config, frame)`

blueyed commented:
So instead of `pytest_enter_pdb` could be kept as-is, but `pytest_debug` added (without adding `pytest_leave_pdb` then)?

RonnyPfannschmidt commented:
@blueyed correct

nicoddemus commented:
Hey guys gentle ping. 😁

nicoddemus commented:
It seems there is still tests failing...

blueyed commented:
Yeah, py27-pexpect still / again(?) failing like before.

blueyed commented:
@nicoddemus 
Ok.
It also needs to be fixed (tests), and I need to review it myself - I am just rebasing it from time to time to see if tests might pass etc.. ;)

nicoddemus commented:
Sure thing, thanks! 😁

RonnyPfannschmidt commented on src/_pytest/hookspec.py:
```diff
@@ -609,3 +609,13 @@ def pytest_enter_pdb(config):
     """
+
+
+def pytest_leave_pdb(config):
```
would it make sense to add the actual debugger as argument here?

blueyed replied:
Good idea.
I'll add a new arg `pdb`.
The same might make sense for `pytest_enter_pdb` then maybe also?!

nicoddemus replied:
Agree to both points. 👍

nicoddemus commented:
Hmm we got a crash on `py36-xdist`:

```
_________________________ testing/code/test_excinfo.py _________________________
[gw0] linux -- Python 3.6.3 /home/travis/build/pytest-dev/pytest/.tox/py36-xdist/bin/python
Worker 'gw0' crashed while running 'testing/code/test_excinfo.py::TestFormattedExcinfo::()::test_repr_traceback_recursion'
```

I will restart that job to see if it is a flaky test.

nicoddemus commented:
Hmm it passed when ran again. Let's keep an eye on this to find out if it was a fluke or actually a flaky test.
